dvc/remote/ssh/__init__.py

```diff
@@ -128,20 +132,37 @@ def ssh(self, path_info):
             password=self.password,
         )
 
-    def exists(self, path_infos):
-        single_path = False
-
-        if not isinstance(path_infos, list):
-            single_path = True
-            path_infos = [path_infos]
+    def exists(self, path_info):
+        with self.ssh(path_info) as ssh:
+            return ssh.exists(path_info.path)
+
+    def batch_exists(self, path_infos, callback):
+        def _exists(chunk_and_channel):
+            chunk, channel = chunk_and_channel
+            ret = []
+            for path in chunk:
+                try:
+                    channel.stat(path)
+                    ret.append(True)
+                except IOError as exc:
+                    if exc.errno != errno.ENOENT:
+                        raise
+                    ret.append(False)
+                callback.update(path)
+            return ret
 
         with self.ssh(path_infos[0]) as ssh:
-            results = [ssh.exists(path_info.path) for path_info in path_infos]
+            with ssh.open_max_sftp_channels() as channels:
```

Comment:
are we reusing the same single connection? what are the limits in terms of how many channels one can open per a connection?

Reply:
The limit of how many SFTP channels we can have per connections is set with the `MaxSessions` option.

It is important to note that the performance increase is not linear, for example, here, the `MaxSessions` is set to `80`, creating a lot of tasks on the remote server:
![2019-06-18-15:21:50](https://user-images.githubusercontent.com/7363250/59717086-eefe8100-91dc-11e9-9378-ea0f9981499f.png)

The performance is worst than when it was set to `40`.

The sweet-spot I found was setting it to `20` (running the whole operation in ~2m, compared to ~3m) when comparing 100K files (it might depend on the load of files)